lxd/instances.go

```diff
@@ -32,9 +33,9 @@ var instancesCmd = APIEndpoint{
 		{Name: "vms", Path: "virtual-machines"},
 	},
 
-	Get:  APIEndpointAction{Handler: instancesGet, AccessHandler: allowProjectPermission("containers", "view")},
-	Post: APIEndpointAction{Handler: instancesPost, AccessHandler: allowProjectPermission("containers", "manage-containers")},
-	Put:  APIEndpointAction{Handler: instancesPut, AccessHandler: allowProjectPermission("containers", "operate-containers")},
+	Get:  APIEndpointAction{Handler: instancesGet, AccessHandler: allProjectsAccessHandler(auth.EntitlementCanView)},
```

Comment:
Kind of. The `allProjectsAccessHandler` checks if you have the `can_view` entitlement on the project (if one is provided). If `all-projects=true` then it checks if you have `can_view` on any projects, if you don't it returns a 403.

Now that I think about it however. Checking `can_view` on a project is incorrect for OpenFGA. We should check `can_view` against all instances. If some instances exist on the server but the user is not allowed to view any of them, then we return 403.

Reply:
So you mean a filtered list or 403 if the filtered result is 0?

Reply:
This is tricky because the underlying drivers are so fundamentally different in approach. There are a few different cases here so I'll try to lay them out. Lets say we have two projects, `project1` and `project2`, and two instances `instance1` and `instance2` in each project respectively:

* RBAC auth. User has `view` permission on `project1` but no permissions for `project2`:
    * `GET /1.0/instances?all-projects=true`: Should list all instances but filter out `instance2`.
    * `GET /1.0/instances?project=project1`: Should list instances in `project1` showing `instance1`.
    * `GET /1.0/instances?project=project2`: Should return 403.
* OpenFGA auth. User has `can_view` entitlement in instance object `instance:project1_instance1` and no other entitlements:
    * `GET /1.0/instances?all-projects=true`: Should list all instances but filter out `instance2`.
    * `GET /1.0/instances?project=project1`: Should list instances in `project1`, filtering out all instances except for `instance1`.
    * `GET /1.0/instances?project=project2`: This is a point of contention. We can either a) return an empty list, b) see that `instance2` exists but we don't have permission to view it, so return a 403, or c) check if we have `can_view` entitlement on the `project2` if we can view `project2` but no instances inside it, return an empty list (and if we can't view `project2`, return a 403).
* RBAC auth. User does not have `view` permission for either project:
    * `GET /1.0/instances?all-projects=true`: Should return 403.
* OpenFGA auth. User does not have `can_view` entitlement on any instances.
    * `GET /1.0/instances?all-projects=true`: Again we can either a) return an empty list, b) see that `instance1` and `instance2` exist but we are not allowed to view them, so return a 403, or c) check if we have `can_view` on any projects, if we don't, return 403, otherwise return an empty list.

Of the three options above I went with c) because it is the most consistent with current behaviour. However it isn't the most OpenFGA-ish way of doing things. Option b) returns a more correct HTTP code but is actually revealing that instances do exist in that project. Option a) is the OpenFGA way but will mean our API will behave differently depending on which authorization driver is loaded.

Reply:
For option A) would it only return an empty list if we have the `can_view` entitlement on project 2, otherwise 403?

Reply:
@monstermunchkin after a brief chat with @tomponline we're going with option a) which is to just filter out instances the user is not able to view. I'll need to make sure the behaviour of RBAC/TLS stays the same however.

Reply:
I had these cases slightly wrong. For RBAC/TLS, the `all-projects` parameter is restricted to admins only. 

I've now added an `rbacPermission` access handler. This returns OK (empty sync response) if neither the RBAC or TLS authorizers are loaded. Otherwise it checks the request context for the user access struct. If the user is an admin then returns OK, else checks the given permission against the project.

Annoyingly this is a leakage of authorization logic outside of the `auth` package but I can't see another way to keep the RBAC/TLS behaviour the same whilst returning empty lists for the OpenFGA driver.

Reply:
I have managed to handle all these cases within the authorization logic now. There is no longer an `rbacPermission` access handler.